Allow decimal degrees angle in rotation popover

OliBomby opened the pull request:
Allows you to make patterns with precise angles and geometry, like a spiral that rotates 22.5 degrees for every circle.

I followed the suggestion by @bdach https://github.com/ppy/osu-framework/pull/6376#issuecomment-2357667053 to use a double bindable setup which bidirectionally linked with value change callbacks.

I let the `Current` bindable also change the `SliderCurrent` bindable so when you create a slider with textbox input with the old api it still functions the same way.

OliBomby committed:
allow different precision between slider and textbox

OliBomby committed:
allow decimal angle in rotation popover

OliBomby committed:
add test

OliBomby committed:
fix precsion

OliBomby committed:
remove public SliderCurrent and input check

bdach commented on osu.Game/Graphics/UserInterfaceV2/SliderWithTextBoxInput.cs:
```diff
@@ -26,10 +27,83 @@ public float KeyboardStep
+        private T? sliderMinValue;
+
+        public T? SliderMinValue
+        {
+            get => sliderMinValue;
+            set
+            {
+                if (value.HasValue && value.Value < current.MinValue)
+                    throw new ArgumentException(@"Minimum value override must be greater than or equal to the current minimum value.");
+
+                sliderMinValue = value;
+                slider.Current = new BindableNumber<T>
+                {
+                    MinValue = value ?? current.MinValue,
+                    MaxValue = sliderMaxValue ?? current.MaxValue,
+                    Default = current.Default,
+                    Precision = sliderPrecision ?? current.Precision,
+                };
+            }
+        }
+
+        private T? sliderMaxValue;
+
+        public T? SliderMaxValue
+        {
+            get => sliderMaxValue;
+            set
+            {
+                if (value.HasValue && value.Value > current.MaxValue)
+                    throw new ArgumentException(@"Maximum value override must be less than or equal to the current maximum value.");
+
+                sliderMaxValue = value;
+                slider.Current = new BindableNumber<T>
+                {
+                    MinValue = sliderMinValue ?? current.MinValue,
+                    MaxValue = value ?? current.MaxValue,
+                    Default = current.Default,
+                    Precision = sliderPrecision ?? current.Precision,
+                };
+            }
```
Why is _this_ exposed? Don't you only need precision mutable? What's the point to this?

OliBomby replied:
Precision is not the only property in `BindableNumber`, so I added these to match the API capability of my previous version. This is also useful to have in the rotation popover because it allows the slider to be constrained to a range while the text input is unconstrained. This allows the user to input angle values beyond the -360 to 360 range which are still well defined and valid. This could be useful when pasting in values from a source with unnormalized values, like an incremental series.

bdach replied:
[-360,360] is plenty. I don't want to allow any more than that. Or, at the *least*, it should be getting normalised back to [-360,360].

OliBomby replied:
Sure normalising it back to that range is an improvement i guess. Still need the extended range in the text box though because the normalization with period 360 is not gonna happen inside `SliderWithTextBoxInput`, and I can think of other uses for extended range.

bdach replied:
Can we just start with the decimal precision thing and nix the extended range for now?

OliBomby replied:
Sure i guess. I'll split it into separate PR.

OliBomby committed:
nix the extended range

OliBomby committed:
retain sliderPrecision when overwriting Current

OliBomby committed:
Avoid updating textbox while typing

If you backspace from '22.5' to '22.' you dont want it to update and remove the dot

bdach reviewed (changes requested):
The tooltip of the slider bar is broken:

![1727162584](https://github.com/user-attachments/assets/ff132618-3ba4-4fea-a545-de82630b7fee)

Maybe worth considering turning it off in this specific usage or something because I don't imagine fixing this is going to be easy otherwise.

OliBomby commented:
> Maybe worth considering turning it off in this specific usage or something because I don't imagine fixing this is going to be easy otherwise.

Yeah the slider bar would need to have the full precision bindable. Maybe revisit https://github.com/ppy/osu-framework/pull/6376 ? We seem to have found a reason to need this.

bdach commented:
> Maybe revisit [ppy/osu-framework#6376](https://github.com/ppy/osu-framework/pull/6376) ? We seem to have found a reason to need this.

Not unless it's better designed. I am not having `SliderBar` have all three of `Step`, `KeyboardStep`, and `Current.Precision`, each of which does a different thing. It's too much.

I'd rather just completely change the precision of the rotation tool to something like 0.1 instead.

OliBomby commented:
> Not unless it's better designed. I am not having `SliderBar` have all three of `Step`, `KeyboardStep`, and `Current.Precision`, each of which does a different thing. It's too much.

I don't see how this is too much. There is no way to make this API more minimal. We'd have all two of `Step` and `KeyboardStep` and I wouldn't count `Current.Precision` as part of the API because its hidden because `SliderBar` doesn't even expose a `BindableNumber`.

Setting the precision on a `SliderBar` by assigning a `BindableNumber` to a `Bindable` is so weird, I think its a better design to not have to use that.

bdach commented:
@ppy/team-client would appreciate additional opinions here because we appear to be at an impasse

peppy commented:
Late response, but the precision coming from `Bindable` is to me a "sane default" and not the way the API should be. I'd probably be okay with having `Step` and `KeyboardStep`.
